modules/contact/include/MechanicalContactConstraint.h

```diff
@@ -131,11 +138,14 @@ class MechanicalContactConstraint : public NodeFaceConstraint
   /// The tolerance of the frictional force for augmented Lagrangian method
   Real _al_frictional_force_tolerance;
 
-  std::set<dof_id_type> _current_contact_state;
+#ifdef LIBMESH_HAVE_PETSC
+  ContactLineSearch * _contact_linesearch;
+#endif
```

Comment:
No... This all needs to be reworked. 

In libMesh, we typically define two versions of objects that have preprocessor dependencies. ContactLineSearch should still be a valid object even when PETSc isn't around, it just won't do much. Trying to do it this way really breaks abstraction. Now you have ifdefs for PETSc spread well outside of the object, you have pointers now instead of variables, etc.

Reply:
I'm all for a better alternative, but my initial thought is that even after bettering the abstraction and removing the `#ifdefs`, I would still like for this to be a pointer. E.g. if the input file doesn't specify `contact` as the line-search, I feel like this should be a null object.

Reply:
OK - that's fine, but it probably shouldn't be a raw pointer. I saw a delete in there which means there's ownership here somewhere. This looks like it should be a smart pointer.